src/middlewared/middlewared/api/v25_04_0/virt.py

```diff
@@ -0,0 +1,279 @@
+from typing import Literal, List, Union, Optional, TypeAlias
+from typing_extensions import Annotated
+
+from pydantic import Field, StringConstraints
+
+from middlewared.api.base import (
+    BaseModel, ForUpdateMetaclass, NonEmptyString,
+    single_argument_args, single_argument_result,
+)
+
+
+class VirtGlobalEntry(BaseModel):
+    id: int
+    pool: str | None = None
+    dataset: str | None = None
+    bridge: str | None = None
+    v4_network: str | None = None
+    v6_network: str | None = None
+    state: str | None = None
+
+
+@single_argument_args('virt_global_update')
+class VirtGlobalUpdateArgs(BaseModel, metaclass=ForUpdateMetaclass):
+    pool: str | None = None
+    bridge: str | None = None
+    v4_network: str | None = None
+    v6_network: str | None = None
+
+
+class VirtGlobalUpdateResult(BaseModel):
+    result: VirtGlobalEntry
+
+
+class VirtGlobalBridgeChoicesArgs(BaseModel):
+    pass
+
+
+class VirtGlobalBridgeChoicesResult(BaseModel):
+    result: dict
+
+
+class VirtGlobalPoolChoicesArgs(BaseModel):
+    pass
+
+
+class VirtGlobalPoolChoicesResult(BaseModel):
+    result: dict
+
+
+class VirtGlobalGetNetworkArgs(BaseModel):
+    name: NonEmptyString
+
+
+@single_argument_result
+class VirtGlobalGetNetworkResult(BaseModel):
+    type: Literal['BRIDGE']
+    managed: bool
+    ipv4_address: NonEmptyString
+    ipv4_nat: bool
+    ipv6_address: NonEmptyString
+    ipv6_nat: bool
+
+
+REMOTE_CHOICES: TypeAlias = Optional[Literal['LINUX_CONTAINERS']]
+
+
+@single_argument_args('virt_instances_image_choices')
+class VirtInstanceImageChoicesArgs(BaseModel):
+    remote: REMOTE_CHOICES = None
+
+
+class ImageChoiceItem(BaseModel):
+    label: str
+    os: str
+    release: str
+    arch: str
+    variant: int
+
+
+class VirtInstanceImageChoicesResult(BaseModel):
+    result: dict[str, ImageChoiceItem]
+
+
+class Device(BaseModel):
+    name: Optional[NonEmptyString] = None
+    dev_type: Literal['USB', 'TPM', 'DISK', 'GPU', 'NIC', 'PROXY']
+    readonly: bool = False
+
+
+class Disk(Device):
+    source: Optional[str] = None
+    destination: Optional[str] = None
+
+
+class NIC(Device):
+    network: NonEmptyString
+
+
+class USB(Device):
+    bus: Optional[int] = None
+    dev: Optional[int] = None
+    product_id: Optional[str] = None
+    vendor_id: Optional[str] = None
+
+
+Proto: TypeAlias = Literal['UDP', 'TCP']
+
+
+class Proxy(Device):
+    source_proto: Proto
+    source_port: int
+    dest_proto: Proto
+    dest_port: int
+
+
+class TPM(Device):
+    path: Optional[str] = None
+    pathrm: Optional[str] = None
+
+
+GPUType: TypeAlias = Literal['PHYSICAL', 'MDEV', 'MIG', 'SRIOV']
+
+
+class GPU(Device):
+    gpu_type: GPUType
+    id: str | None = None
+    gid: Optional[int] = None
+    uid: Optional[int] = None
+    mode: Optional[int] = None
```

Comment:
Can we more precisely define uid, gid, and mode?

Reply:
Do you mean restricting gid/uid to 32 bits and mode to 9 bits?